expression/builtin_time.go

```diff
@@ -3280,1581 +3388,616 @@ func (du *baseDateArithmetical) vecGetIntervalFromReal(b *baseBuiltinFunc, input
+}
+
+func getIntervalFromInt(da *baseDateArithmetical, ctx sessionctx.Context, args []Expression, row chunk.Row, unit string) (string, bool, error) {
+	return da.getIntervalFromInt(ctx, args, row, unit)
+}
+
+func getIntervalFromReal(da *baseDateArithmetical, ctx sessionctx.Context, args []Expression, row chunk.Row, unit string) (string, bool, error) {
+	return da.getIntervalFromReal(ctx, args, row, unit)
+}
+
+func getIntervalFromDecimal(da *baseDateArithmetical, ctx sessionctx.Context, args []Expression, row chunk.Row, unit string) (string, bool, error) {
+	return da.getIntervalFromDecimal(ctx, args, row, unit)
+}
+
+type funcVecGetIntervalForDateAddSub func(da *baseDateArithmetical, b *baseBuiltinFunc, input *chunk.Chunk, unit string, result *chunk.Column) error
+
+func vecGetIntervalFromString(da *baseDateArithmetical, b *baseBuiltinFunc, input *chunk.Chunk, unit string, result *chunk.Column) error {
+	return da.vecGetIntervalFromString(b, input, unit, result)
+}
+
+func vecGetIntervalFromInt(da *baseDateArithmetical, b *baseBuiltinFunc, input *chunk.Chunk, unit string, result *chunk.Column) error {
+	return da.vecGetIntervalFromInt(b, input, unit, result)
+}
+
+func vecGetIntervalFromReal(da *baseDateArithmetical, b *baseBuiltinFunc, input *chunk.Chunk, unit string, result *chunk.Column) error {
+	return da.vecGetIntervalFromReal(b, input, unit, result)
+}
+
+func vecGetIntervalFromDecimal(da *baseDateArithmetical, b *baseBuiltinFunc, input *chunk.Chunk, unit string, result *chunk.Column) error {
+	return da.vecGetIntervalFromDecimal(b, input, unit, result)
+}
+
+type addSubDateFunctionClass struct {
 	baseFunctionClass
+	timeOp      funcTimeOpForDateAddSub
+	durationOp  funcDurationOpForDateAddSub
+	setPbCodeOp funcSetPbCodeOp
 }
 
-func (c *addDateFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (sig builtinFunc, err error) {
+func (c *addSubDateFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (sig builtinFunc, err error) {
 	if err = c.verifyArgs(args); err != nil {
 		return nil, err
 	}
 
 	dateEvalTp := args[0].GetType().EvalType()
-	if dateEvalTp != types.ETString && dateEvalTp != types.ETInt && dateEvalTp != types.ETDuration {
+	// Some special evaluation type treatment.
+	// Note that it could be more elegant if we always evaluate datetime for int, real, decimal and string, by leveraging existing implicit casts.
+	// However, MySQL has a weird behavior for date_add(string, ...), whose result depends on the content of the first argument.
+	// E.g., date_add('2000-01-02 00:00:00', interval 1 day) evaluates to '2021-01-03 00:00:00' (which is normal),
+	// whereas date_add('2000-01-02', interval 1 day) evaluates to '2000-01-03' instead of '2021-01-03 00:00:00'.
+	// This requires a customized parsing of the content of the first argument, by recognizing if it is a pure date format or contains HMS part.
+	// So implicit casts are not viable here.
+	if dateEvalTp == types.ETTimestamp {
```

Comment:
Are there any tests with Timestamps using different time zones (including Daylight Saving Time)?
If not I would suggest having something for Europe/Amsterdam like
timestamp column = '2022-03-27 01:30:00' (CET / GMT+1)
DATE_ADD(ts_col, INTERVAL 1 HOUR) => '2022-03-27 02:30:00' Datetime (non-existing CE(S)T timestamp)
timestamp column = '2022-10-30 02:30:00' (CEST / GMT+2)
DATE_ADD(ts_col, INTERVAL 1 HOUR) => '2022-10-30 03:30:00' Datetime (Two hours later if interpreted as CET timestamp, GMT+1)
timestamp column = '2022-10-30 02:30:00' (CET / GMT+1)
DATE_ADD(ts_col, INTERVAL 1 HOUR) => '2022-10-30 03:30:00' Datetime

Reply:
I'll do some experiments and come to this topic later.

Reply:
I did this in MySQL:
```
set global time_zone = 'cet';
create table t1(ts timestamp);
insert into t1 values ('2022-03-27 01:30:00'), ('2022-10-30 02:30:00');
select date_add(ts, interval 1 hour) from t1;
+-------------------------------+
| date_add(ts, interval 1 hour) |
+-------------------------------+
| 2022-03-27 02:30:00           |
| 2022-10-30 03:30:00           |
+-------------------------------+
```

I think this meets my expectation. Did I get this right?

Reply:
Tried the same thing on TiDB (with this PR) and obtained the same result as MySQL.

If the experiment itself is correct, I'll move forward to add these cases in code.

Reply:
Case `TestDateAddForNonExistingTimestamp` in `integration_tests.go` added.

Reply:
Thanks for adding the test, I did a minor suggestion in a new review.